Connect to a new outbound peer if our tip is stale

sdaftuar opened the pull request:
This is an alternative approach to #11534.  Rather than disconnect an outbound peer when our tip looks stale, instead try to connect to an additional outbound peer.

Periodically, check to see if we have more outbound peers than we target (ie if any extra peers are in use), and if so, disconnect the one that least recently announced a new block (breaking ties by choosing the newest peer that we connected to).

sdaftuar committed:
net: Allow connecting to extra outbound peers

sdaftuar committed:
Track tip update time and last new block announcement from each peer

promag committed:
Add CConnmanTest to mutate g_connman in tests

sdaftuar committed:
Connect to an extra outbound peer if our tip is stale

If our tip hasn't updated in a while, that may be because our peers are
not relaying blocks to us that we would consider valid. Allow connection
to an additional outbound peer in that circumstance.

Also, periodically check to see if we are exceeding our target number of
outbound peers, and disconnect the one which has least recently
announced a new block to us (choosing the newest such peer in the case
of tie).

sdaftuar commented:
I'm still working on a unit test for this, but wanted to get some eyes on it for consideration in 0.15.0.2.

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -2928,10 +2940,61 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+        // us a new block, with ties broken by choosing the more recent
+        // connection (higher node id)
+        NodeId worst_peer = -1;
+        int64_t oldest_block_announcement = time_in_seconds;
```
I'd prefer to see this be a int_64 max, so I don't need to reason about how monotone our clocks aren't.

sdaftuar replied:
Done

gmaxwell commented:
Remind me how the scheduler works... when blocks are slow is this going to result in all hosts on the network triggering their extra peer at basically the same time?  Or do we not need to insert a bit of randomness in order to spread them out over the checking interval?

sdaftuar committed:
Add unit test for stale tip checking

sdaftuar commented:
Regarding the scheduler, I mentioned this on IRC to @gmaxwell:
> gmaxwell: my understanding of the scheduler is that it'll start scheduling callbacks at some point after startup, so i wouldn't expect the network to be completely synced
> gmaxwell: also there's random drift, since the scheduler schedules the next callback N milliseconds after the prior one finishes 
> but worth discussing whether the interval is short enough that there still might be too much concentration of everyone trying to find a new peer

I updated this PR with a unit test.

sdaftuar commented:
Needed rebase after #11490.

TheBlueMatt commented on src/net.h:
```diff
@@ -251,6 +253,16 @@ class CConnman
+
+    // Return the number of outbound peers we have in excess of our target (eg,
+    // if we previously called SetTryNewOutboundPeer(true), and have since set
+    // to false, we may have extra peers that we wish to disconnect).
```
You may want to note the (IMO) critical part of the cpp comment:
"Also exclude peers that haven't finished initial connection handshake yet" ie "this may return a value less than the number of outbound connections - the number of normal outbound connection slots in cases where some outbound connections are not yet fully connected".

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1068,7 +1068,7 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     CAddress addr;
     int nInbound = 0;
-    int nMaxInbound = nMaxConnections - (nMaxOutbound + nMaxFeeler);
+    int nMaxInbound = nMaxConnections - (nMaxOutbound + nMaxFeeler + nMaxExtraOutbound);
```
I'm confused - it looks like we wont ever have both a feeler and an extra outbound open at one time, why change this?

sdaftuar replied:
> it looks like we wont ever have both a feeler and an extra outbound at one time

We will in general have both a feeler and an extra outbound at the same time.  There was an incorrect comment (from a first iteration of this patch) describing the extra outbound peer as "stealing" the feeler connection, but that is not actually true.  I'll delete that comment.

TheBlueMatt reviewed (commented):
Didnt get very far yet, but given deadlines, figured I'd post early and often :p.

sdaftuar commented on src/net.h:
```diff
@@ -413,6 +426,10 @@ class CConnman
     std::thread threadOpenConnections;
     std::thread threadMessageHandler;
+
+    /** flag for stealing a feeler connection to find a new outbound peer */
```
whoops, this comment is wrong, will fix.

sdaftuar commented:
Pushed up a few small fixups.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -2941,6 +2953,57 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+                // Only disconnect a peer that has been connected to us for
+                // some reasonable fraction of our check-frequency, to give
+                // it time for new information to have arrived.
+                if (time_in_seconds - pnode->nTimeConnected > STALE_CHECK_INTERVAL/2) {
```
Shouldnt this almost always be true (because we only call this function once every STALE_CHECK_INTERVAL)? Also, I think we can turn this down way more than 5 minutes, if we think we're behind, and we connect to someone, it shouldn't take them 5 minutes to send us a header, and if it does, we probably weren't behind and our existing peers are fine, so we should just disconnect them.

sdaftuar replied:
Maybe almost always, but definitely not always -- our outbound peers can initiate disconnects, too, and it might take us a while to find one.

I've pushed up changes to split the peer eviction onto a shorter timer (2.5 minutes), where we only require 30 seconds of connectivity to disconnect.

TheBlueMatt reviewed (commented):
Didnt review test changes yet.

sdaftuar commented:
@TheBlueMatt raised concerns about holding the 9th connection for too long.  One concern would be if it we are reducing the number of available inbound slots on the network by too much. Another potentially more serious concern is that if we're holding the 9th connection long enough for a new block to be found and relayed, then over time we'll be selecting for the fastest connectivity.

In the extreme, you could imagine that this algorithm would reduce to nodes grinding away, trying to connect to the peer that was closest to the miners (or the FIBRE network or similar), which is probably not what we want.

In this case, we only intend for the 9th connection to tell us whether we're really behind (eg because our existing peers are all broken) or if the network is just slow.  We can determine that very quickly after connection, so I think I'm going to change this so that we disconnect extra peers more quickly (ie on a faster timer than every 10 minutes, and requiring a shorter connection time), and reset the TryExtraOutboundPeer flag in CConnman after disconnecting an extra peer, essentially waiting until we decide at the next stale-tip-check that we're still stale, before trying another 9th peer.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2382,6 +2399,10 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         // are still present, however, as belt-and-suspenders.
 
+        if (received_new_header && pindexLast->nChainWork > chainActive.Tip()->nChainWork) {
+            nodestate->m_last_block_announcement = GetTime();
```
oops, yes.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -436,6 +436,15 @@ void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman* connman) {
+    return g_last_tip_update < GetTime() - consensusParams.nPowTargetSpacing * 3 && mapBlocksInFlight.empty();
+}
+
+// Requires cs_main
```
Sounds good

sdaftuar commented on src/net.cpp:
```diff
@@ -1781,9 +1817,14 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+
+        int desired_outbound = nMaxOutbound + (extra_outbound ? nMaxExtraOutbound : 0);
+
+        if (nOutbound >= desired_outbound) {
```
Not sure if this comment was made before we chatted offline, but it is possible to have both a feeler and an "extra" outbound peer simultaneously (just not in the same loop iteration, but that's nonsensical anyway since we only initiate a single connection in a single loop iteration, regardless).

sdaftuar commented:
Addressed @TheBlueMatt's nits.

theuni commented on src/net.cpp:
```diff
@@ -1693,6 +1693,38 @@ void CConnman::ProcessOneShot()
+// Also exclude peers that haven't finished initial connection handshake yet
+// (so that we don't decide we're over our desired connection limit, and then
+// evict some peer that has finished the handshake)
+int CConnman::GetExtraOutboundCount()
```
I think this might be greatly simplified by extending CSemaphore to have an adjustable threshold.

sdaftuar replied:
Any thoughts on @TheBlueMatt's suggestion (mentioned on IRC yesterday) to just use the feeler connection for the extra peer, rather than initiate a potential 10th connection?

https://botbot.me/freenode/bitcoin-core-dev/2017-10-27/?msg=92828477&page=2

I'm leaning towards doing that, so nMaxExtraOutbound would go away.  I guess this function wouldn't change though.

sipa commented:
Needs rebase.

sdaftuar commented:
Squashed and rebased.

theuni commented:
Without seeing the code, I think that would probably be simpler and more
obvious.

I wonder if we could actually combine the behavior with feelers. Launch a
feeler every 20 min or so as usual, and rather than disconnecting
immediately, replace a worse connection if necessary, and upgrade from
feeler to full outbound.

On Oct 28, 2017 11:11 AM, "Suhas Daftuar" <notifications@github.com> wrote:

> *@sdaftuar* commented on this pull request.
> ------------------------------
>
> In src/net.cpp
> <https://github.com/bitcoin/bitcoin/pull/11560#discussion_r147556553>:
>
> > +    return m_try_another_outbound_peer;
> +}
> +
> +void CConnman::SetTryNewOutboundPeer(bool flag)
> +{
> +    LOCK(m_cs_outbound_peer);
> +    m_try_another_outbound_peer = flag;
> +}
> +
> +// Return the number of peers we have over our outbound connection limit
> +// Exclude peers that are marked for disconnect, or are going to be
> +// disconnected soon (eg one-shots and feelers)
> +// Also exclude peers that haven't finished initial connection handshake yet
> +// (so that we don't decide we're over our desired connection limit, and then
> +// evict some peer that has finished the handshake)
> +int CConnman::GetExtraOutboundCount()
>
> Any thoughts on @TheBlueMatt <https://github.com/thebluematt>'s
> suggestion (mentioned on IRC yesterday) to just use the feeler connection
> for the extra peer, rather than initiate a potential 10th connection?
>
> https://botbot.me/freenode/bitcoin-core-dev/2017-10-27/?
> msg=92828477&page=2
>
> I'm leaning towards doing that, so nMaxExtraOutbound would go away. I
> guess this function wouldn't change though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/11560#discussion_r147556553>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZdEzhjrlXqpR_iPo_-CIcNtvbYKR8cks5sw0QdgaJpZM4QGjZX>
> .
>

sdaftuar commented:
Fixed a bug and changed the extra peer connection to be exclusive with a feeler, rather than in addition to a feeler.

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3021,81 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+        } else {
+            connman->SetTryNewOutboundPeer(false);
+        }
+        m_stale_tip_check_time += STALE_CHECK_INTERVAL;
```
If this loop fails to run when it should-- e.g. host it suspended, then later when it does run it will run additional times.  I believe this should be = time_in_seconds + STALE_CHECK_INTERVAL.

gmaxwell replied:
Doubly so because nothing appears to initialize it to a time to begin with...

sdaftuar replied:
Thanks for catching that; fixed.

sdaftuar commented on src/net.cpp:
```diff
@@ -1781,7 +1813,8 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
         bool fFeeler = false;
-        if (nOutbound >= nMaxOutbound) {
+
+        if (nOutbound >= nMaxOutbound && !GetTryNewOutboundPeer()) {
```
Yes, I think if `GetTryNewOutboundPeer` is set, then we keep trying outbound peers up to the semaphore count, at which point I believe we block.  Is that problematic for some reason?  I'm not super familiar with the design here but my understanding from reading the code was that this was how this logic was originally written (ie before feelers were introduced, with the `continue` in the `if` block to avoid consuming the last semaphore value via a non-feeler).

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1781,7 +1813,8 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
         bool fFeeler = false;
-        if (nOutbound >= nMaxOutbound) {
+
+        if (nOutbound >= nMaxOutbound && !GetTryNewOutboundPeer()) {
```
Grrr, indeed, I suppose this is correct.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -2971,6 +3030,84 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+                // it time for new information to have arrived.
+                // Also don't disconnect any peer we're trying to download a
+                // block from.
+                constexpr int64_t min_connect_time = EXTRA_PEER_CHECK_INTERVAL / 5;
```
Why check if we need to disconnect every 2.5 minutes and only require it being online 30 seconds? Why not just make both 30 seconds (or maybe 45 and 30) so that we dont always end up holding our extra peer for 2.5 minutes?

sdaftuar replied:
I'm just not really sure how frequent is too-frequent for scheduled callbacks; there's a tradeoff between wasting time on calls (especially in this framework, where we always have the callback fire to check for eviction, even if we've never called SetExtraOutboundPeer()) and gaining precision with how long we allow the extra peer to be connected, and it's not clear to me how to figure out the optimal point....

TheBlueMatt replied:
The callback in the connman->GetExtraOutboundCount() == 0 case should be almost free, no?

sdaftuar replied:
Alright, changed to 45 seconds.

TheBlueMatt reviewed (commented):
utACK ff5f5cd2b9ca9c32200ddb85a7d4510c334576b4 (minus test changes and I still prefer a shorter check interval).

jimpo commented on src/net_processing.h:
```diff
@@ -55,6 +60,10 @@ class PeerLogicValidation : public CValidationInterface, public NetEventsInterfa
+    void CheckForStaleTipAndEvictPeers(const Consensus::Params &consensusParams);
+    void EvictExtraOutboundPeers(int64_t time_in_seconds);
+
+    int64_t m_stale_tip_check_time;
```
Should this be private?

sdaftuar replied:
Sounds good, done.

sdaftuar commented:
Updated to address a nit from @jimpo, and reduce the check-for-eviction interval down to 45 seconds.

Squashed https://github.com/sdaftuar/bitcoin/commits/11560.0 -> 8a27fdf9ae691eb32b8b5e9f531e16266f8227a8

achow101 commented on src/net_processing.cpp:
```diff
@@ -427,6 +435,15 @@ void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman* connman) {
+    if (g_last_tip_update == 0) {
+        g_last_tip_update = GetTime();
+    }
+    return g_last_tip_update < GetTime() - consensusParams.nPowTargetSpacing * 3 && mapBlocksInFlight.empty();
```
Since it is not uncommon for blocks to take more than 30 minutes to be found, maybe this threshold should be higher? It also seems like every node is going to be making that outbound connection at around the same time, maybe there should be some element of randomness to that to avoid some sort of fingerprinting?

sdaftuar replied:
I think the scheduler behavior will mean that nodes are going to spread out their extra connections over a 10 minute window.  Since we already do feeler connections on a (poisson-random) 2 minute interval, and this is taking the place of a feeler, I don't think this will introduce too much network load.

If there's a fingerprinting concern, I can try to come up with some fuzzing on when the connection is made (maybe another poisson-random fuzz?), but I'm not sure I understand the specific fingerprinting risk with this logic?

achow101 replied:
> but I'm not sure I understand the specific fingerprinting risk with this logic?

I don't think there is any. I had a fingerprinting concern but then realized that that concern was stupid and stopped being concerned about it.

TheBlueMatt commented:
re-utACK 8a27fdf9ae691eb32b8b5e9f531e16266f8227a8

achow101 commented:
utACK 8a27fdf9ae691eb32b8b5e9f531e16266f8227a8

theuni commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3021,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+
+        LOCK(cs_main);
+
+        connman->ForEachNode([&](CNode* pnode) {
```
Can we loop through mapNodeState here instead? I'd really like to avoid gluing these layers back together.

Also, I believe that would mean that the ```nBlocksInFlight != 0``` check could be applied here.

sdaftuar replied:
I just wrote this in my other response, but I don't think the blocks in flight check is a good thing to sort on -- I think there may be edge cases where somehow your slowest peer gets protected, and you would evict someone else.

I'll see what it looks like when I try to loop mapNodeState instead.  I think I'll still have to call `ForNode` inside that loop in order to get to the pnode, so I don't really see the difference in design...?  Can you explain a little more what the design goal is for stuff like this?

theuni commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3021,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+                // Also don't disconnect any peer we're trying to download a
+                // block from.
+                CNodeState &state = *State(pnode->GetId());
+                if (time_in_seconds - pnode->nTimeConnected > MINIMUM_CONNECT_TIME && state.nBlocksInFlight == 0) {
```
Doesn't this mean that if the worst peer is one second worse than the second-worst, and the worst has a block in flight, both would stay connected? In that case, I would expect the second-worst to be evicted.

Adding to the comment above, I think that checking for nBlocksInFlight in the worst_peer calculation above avoids this?

sdaftuar replied:
I don't think it's necessary to evict in that situation, since we'll just try to evict again in 45 seconds, and I thought it was more conservative to not evict.  For instance, the block in flight might be at equal work to our tip, rather than more work, and we wouldn't want to disconnect the older peer in that situation where we're just downloading an equal-work-chain-tip.

Also if you evict the second-worst peer while a block download is in flight, then it's possible that the new peer will stall your block download, and you'll have lost one of your earlier peers unnecessarily.  

Anyway I mostly just want to avoid anything that can be gamed, and postponing eviction seems like a safer option than choosing a different peer to evict, I think?

theuni commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3021,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+
+    int64_t time_in_seconds = GetTime();
+
+    EvictExtraOutboundPeers(time_in_seconds);
```
Why is this done on a timer rather than making InitializeNode aware that a new node is a replacement?

sdaftuar replied:
I think it'd be difficult to try to do in InitializeNode, because you don't know who you are going to want to evict until you're ready to do the eviction.  For example, what happens if one (or all) of your first 8 peers disconnects in between the InitializeNode for your 9th peer, and eviction time?

theuni replied:
ok, understood

sdaftuar commented:
@theuni Pushed a commit that rewrites the `ForEachNode()` into a loop on `mapNodeState`.

theuni commented on src/net_processing.h:
```diff
@@ -27,13 +28,19 @@ static constexpr int64_t HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER = 1000; // 1ms/head
+/** How frequently to check for extra outbound peers and disconnect, in seconds */
+static constexpr int64_t EXTRA_PEER_CHECK_INTERVAL = 45;
+/** Minimum time an outbound-peer-eviction candidate must be connected for, in order to evict, in seconds */
+static constexpr int64_t MINIMUM_CONNECT_TIME = 30;
```
I think we should probably bump this to 60sec, to match the minimum time allowed for the handshake. Or was 30sec a deliberate choice?

That would also give us the handy property that all fSuccessfullyConnected nodes can be disconnected without checking the connection duration.

gmaxwell replied:
>  disconnected without checking the connection duration.

Huh.. then you'd disconnect someone who completed the handshake fast, without giving them more than a few seconds to transfer blocks.

theuni replied:
Sorry, I meant that in the context of the check here: https://github.com/bitcoin/bitcoin/pull/11560/commits/21e8b03a353abd0f2f4190194f224ad92d6f5bc8#diff-eff7adeaec73a769788bb78858815c91R3055

TheBlueMatt replied:
I dont really think it matters here - these are separate, we're seeking a new useful outbound peer, if it doesn't respond in 30 seconds with a headers message, it probably isnt/wont be useful. Honestly I think we should significantly decrease the headers handshake timeout, there isn't really a scenario where you should need 60 seconds for a few packets back and forth, even in super lossy environments.

sdaftuar replied:
ping @TheBlueMatt

gmaxwell commented:
ACK, I've had various versions of this running on a mainnet node for about 5 days now, but with the stall timeout cut down to a fraction of the default value so I could see it in action.  Latest versions appear to work as expected.

One gotcha is that the Turning off message will pretty much never print and is kind of pointless and confusing.  Perhaps it should be moved to the inside of settrynewoutbound.

sdaftuar commented:
> I'm not too worried about matt's unbalancing point: this only runs fairly rarely, it runs fairly slowly, and there are 4 protected peers that it won't unbalance even in the worst case.

Actually, the protected-peers logic from the chain sync eviction algorithm wasn't being used here.  After further thought I don't actually have any good reason not to protect those peers from eviction, and I think the same reasons to protect peers apply here, so I've updated this PR accordingly.

I also pushed up a commit that improves the logging.

theuni commented:
utACK as-is, but I'd really like to re-work this in master post-merge. We can separate the net vs net_processing states so that they're not so intertwined.

sdaftuar commented:
Squashed https://github.com/sdaftuar/bitcoin/commits/11560.1 -> dac71cdba95a1b004d2daae322fc026f184ea082

sipa commented on src/net.h:
```diff
@@ -413,6 +426,12 @@ class CConnman
+    /** flag for deciding to connect to an extra outbound peer,
+     *  in excess of nMaxOutbound
+     *  This takes the place of a feeler connection */
+    CCriticalSection m_cs_outbound_peer;
```
If this CS doesn't protect more than a single boolean, it's better to instead just use an `std::atomic<bool>`.

sdaftuar replied:
Done

sdaftuar commented:
@TheBlueMatt just pointed out to me that ForNode has to walk the vector of CNode's to find the one we're looking for, so by rewriting the ForEachNode to a loop on mapNodeState (as suggested by @theuni up here: https://github.com/bitcoin/bitcoin/pull/11560#discussion_r148040947), we're now doing n^2 work instead of n log n work...  That seems unfortunate to me -- default configurations for listening nodes have > 100 connections right?  

I'd like to revert that change.

sdaftuar commented:
Pushed a commit (to be squashed) which reverts the loop back to a ForEachNode, rather than walking mapNodeState.  Also added a nullptr check on the lookup into mapNodeState, just to be safe.

TheBlueMatt commented:
re-utACK 79f686b7344b9612123fddec8397a01ab0356713

sdaftuar commented:
Squashed https://github.com/sdaftuar/bitcoin/commits/11560.3 -> 7d9415396b049dd1477febb71930a70870c5a541

theuni commented:
utACK 7d94153.

gmaxwell commented:
re-ut part of my ACK. I'll move my tests onto this code.

sipa commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3022,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+
+void PeerLogicValidation::CheckForStaleTipAndEvictPeers(const Consensus::Params &consensusParams)
+{
+    if (connman == nullptr) return;
```
Nit: this is a slightly easier to reason about if you make the member definition for connman `CConnman* const connman;` (which compiles without other changes)

sdaftuar replied:
Sounds good.

sipa commented on src/net_processing.cpp:
```diff
@@ -3004,6 +3022,83 @@ void PeerLogicValidation::ConsiderEviction(CNode *pto, int64_t time_in_seconds)
+            // Ignore non-outbound peers, or nodes marked for disconnect already
+            if (!IsOutboundDisconnectionCandidate(pnode) || pnode->fDisconnect) return;
+            CNodeState *state = State(pnode->GetId());
+            if (state == nullptr) return; // shouldn't be possible, but just in case
```
Make assert?

sdaftuar replied:
I thought our preference was to not assert except in situations where node shutdown is necessary to avoid eg consensus failure or lost funds or something similarly bad?  I'll add the assert if that's your preference, but I was just nervous about introducing a crash if I got some race condition wrong.

I could add a LogPrinf() here instead perhaps?

TheBlueMatt replied:
Agree with @sdaftuar, not only should we not be asserting here, but I want to move away from net_processing relying on specific order-of-operations for deletion of CNode/FinalizeNode in net.

sipa replied:
@TheBlueMatt Perhaps asserting is too harsh, but silently ignoring the fact that an underlying assumption is broken seems bad too. 

Also, it seems very reasonable to assume that net does not invoke any net_processing operations before initializing a node, or after finalizing it? That seems easy to guarantee, and very hard for handlers to deal with if untrue.

TheBlueMatt replied:
@sipa If we moved mapNodeState outside of cs_main (which we should do) plus take #10697, then this would be a legitimate race that doesn't matter - if FinalizeNode is called while this loop is running, we may hit this condition and evict our 8th peer because our 9th just disconnected, which may happen either way because the peer may disconnect a millisecond after we evict. This is a perfectly fine thing that isn't a real issue, but re-introducing the strong assumption about when things are removed from vNodes vs when FinalizeNode is called is needless and annoying.

sipa replied:
Got it, I withdraw my comment.

sipa reviewed (commented):
utACK 7d9415396b049dd1477febb71930a70870c5a541

I did not review the test changes.

sdaftuar commented:
Squashed https://github.com/sdaftuar/bitcoin/commits/11560.4 -> 626291508c433488439b662f2e88882048fb59fb

```
$ git diff 626291508c433488439b662f2e88882048fb59fb 7d9415396b049dd1477febb71930a70870c5a541
diff --git a/src/net_processing.h b/src/net_processing.h
index 0a49972eed..c1f1274f80 100644
--- a/src/net_processing.h
+++ b/src/net_processing.h
@@ -37,7 +37,7 @@ static constexpr int64_t MINIMUM_CONNECT_TIME = 30;
 
 class PeerLogicValidation : public CValidationInterface, public NetEventsInterface {
 private:
-    CConnman* const connman;
+    CConnman* connman;
 
 public:
     explicit PeerLogicValidation(CConnman* connman, CScheduler &scheduler);
```

gmaxwell commented:
re-ACK.

theuni commented:
utACK 626291508c433488439b662f2e88882048fb59fb

MyS5Droid commented:
I had a slow sync issue which may have been related to and/or fixed by this... See https://github.com/bitcoin/bitcoin/issues/12514

I eventually did get the recompile of the GitHub code but the version # did not change.

I hope my issue/resolution helps in fixing this issue or confirming it is fixed :)
